### PR TITLE
James 3763 BlobStore backed MailRepository

### PR DESCRIPTION
**Pull request by mbaechler**
This branch includes commits from https://github.com/apache/james-project/pull/985  and https://github.com/apache/james-project/pull/986 and add the blobstore implementation on top.

### CODE REVIEW

File: server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  @throws[MessagingException]
+  override def store(mc: Mail): MailKey = {
+    mimeMessageStore.save(mc.getMessage)
+      .flatMap(mimePartsId => mailMetadataStore.save((mc, mimePartsId)))
+      .map(mailPartsId => mailPartsId.toMailKey)
+      .block()
+  }
+
+  @throws[MessagingException]
+  override def size: Long = Flux.from(blobStore.listBlobs(bucketName)).count().block()
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala:139]: There might be a higher level S3 method for getting bucket size, that might be more efficient? Would we benefit from adding it to the blobStore API ?

**Review by chibenwa — COMMENTED** [review]
Any chance to get this plugged into the distributed server? Documented?

**Comment by chibenwa** [comment]
You are missing ticket numbers on commits.

File: server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala (hunk 1 above)
Comment (mbaechler) [reply server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala:139]: I could not find one

**Comment by mbaechler** [comment]
> Any chance to get this plugged into the distributed server? Documented?

We are not going to deploy this distribution for now, we'll probably only integrate it to the smtp distribution we want to deploy.

It will need some documentation efforts that we'll handle.

Feel free to reuse this impl in other contexts

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
```
23:41:27,611 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (check-style) on project blob-common: You have 2 Checkstyle violations. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (check-style) on project blob-common: You have 2 Checkstyle violations.
```

Cherry pick of the checkstyle fixups?

**Comment by jeantil** [comment]
Won't it be easier to merge the other 2 PRs then rebase this one ?

File: server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala [hunk 2]
```diff
@@ -0,0 +1,186 @@
+package org.apache.james.mailrepository.blob
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala:1]: License

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**